### PR TITLE
Create and get communities.

### LINKED ISSUE #47: [BACKEND] Implement API to create community.

### LINKED ISSUE #48: [BACKEND] Implement API to get user's communities.

### LINKED ISSUE #57: [BACKEND] Implement API to get a community's details.

### PR DESCRIPTION
**Pull request by git-adithyanair**

### CODE REVIEW

File: backend/Makefile [hunk 1]
```diff
@@ -16,5 +16,11 @@ migratedown:
 sqlc:
 	sqlc generate
 
+devup:
+	docker compose up
+
+devdown:
+	docker compose down && docker rmi backend-good-grocer-api
+
```
Comment (yililiu8) [inline backend/Makefile:24]: niceee

File: backend/api/community_create.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	YCoord  float64 `json:"y_coord" binding:"required,numeric"`
+	PlaceID string  `json:"place_id" binding:"required"`
+	Address string  `json:"address" binding:"required"`
+}
+
+type CreateCommunityRequest struct {
+	Name         string                        `json:"name" binding:"required"`
+	CenterXCoord float64                       `json:"center_x_coord" binding:"required,numeric"`
+	CenterYCoord float64                       `json:"center_y_coord" binding:"required,numeric"`
+	Range        int32                         `json:"range" binding:"required,numeric"`
```
Comment (yililiu8) [inline backend/api/community_create.go:25]: Should this be required? I was assuming we'd have a default value for this and if for some reason the admin wanted to change it they could.

File: backend/api/community_create.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+				CenterXCoord: req.CenterXCoord,
+				CenterYCoord: req.CenterYCoord,
+				Range:        req.Range,
+				PlaceID:      req.PlaceID,
+				Address:      req.Address,
+			}
+
+			community, err := server.queries.CreateCommunity(ctx, arg)
+			if err != nil {
+				if err == sql.ErrNoRows {
```
Comment (yililiu8) [inline backend/api/community_create.go]: Is this error possible when creating a row?

File: backend/api/user_communities.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	"database/sql"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/git-adithyanair/cs130-group-project/token"
+)
+
+func (server *Server) GetUserCommunities(ctx *gin.Context) {
+
+	authPayload := ctx.MustGet(authorizationPayloadKey).(*token.Payload)
```
Comment (yililiu8) [inline backend/api/user_communities.go:13]: Check that authPayload != nil here before using?

File: backend/api/user_communities.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+	communities, err := server.queries.GetUserCommunities(ctx, authPayload.UserID)
+	if err != nil {
+		if err != sql.ErrNoRows {
+			ctx.JSON(http.StatusInternalServerError, errorResponse(err))
+		}
+		return
```
Comment (yililiu8) [inline backend/api/user_communities.go]: Why do we not have to give a response before the return if there is an error that isn't `ErrNoRows`?

File: backend/Makefile (hunk 1 above)
Comment (git-adithyanair) [reply backend/Makefile:24]: heheh making life easier one step at a time

File: backend/api/community_create.go (hunk 2 above)
Comment (git-adithyanair) [reply backend/api/community_create.go:25]: I think when they're creating a community they should provide one, and if not the frontend can select a default value

File: backend/api/user_communities.go (hunk 4 above)
Comment (git-adithyanair) [reply backend/api/user_communities.go:13]: authPayload can never be nil, check the middleware.go file

File: backend/api/community_create.go (hunk 3 above)
Comment (git-adithyanair) [reply backend/api/community_create.go]: great point lmaooo

File: backend/api/user_communities.go (hunk 5 above)
Comment (git-adithyanair) [reply backend/api/user_communities.go]: if there is some error and it is not the ErrNoRows, then something went wrong and we need to return a 500. if the error was ErrNoRows then communities will just be an empty list and we send that in the 200 response

Comment (yililiu8) [reply backend/api/user_communities.go]: wait but doesn't the code return before reaching the response on line 23 if `err == sql.ErrNoRows`? If it returns without `ctx.JSON...` does that automatically mean its sending 200 with an empty list?

Comment (git-adithyanair) [reply backend/api/user_communities.go]: ohh yes good point I didn't catch that

**Review by yililiu8 — APPROVED** [review]

**Pull request merged**